File: python/kruskals_algorithm.py
class DSU:
    def __init__(self, nodes):
        self.parent = {n: n for n in nodes}
        self.rank = {n: 0 for n in nodes}

    def find(self, i):
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    def union(self, i, j):
        root_i = self.find(i)
        root_j = self.find(j)

        if root_i != root_j:
            if self.rank[root_i] < self.rank[root_j]:
                self.parent[root_i] = root_j
            elif self.rank[root_i] > self.rank[root_j]:
                self.parent[root_j] = root_i
            else:
                self.parent[root_j] = root_i
                self.rank[root_i] += 1
            return True
        return False

def kruskals_mst(nodes, edges):
    
    # Edges are (weight, u, v)
    edges.sort()
    
    dsu = DSU(nodes)
    mst = []
    
    for w, u, v in edges:
        if dsu.union(u, v):
            mst.append((u, v, w))

    return mst

# Example
# nodes: a list of all vertices
N = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I']
# edges: a list of (weight, u, v)
E = [
    (4, 'A', 'B'), (8, 'A', 'H'),
    (8, 'B', 'C'), (11, 'B', 'H'),
    (7, 'C', 'D'), (4, 'C', 'F'),
    (9, 'D', 'E'), (14, 'D', 'F'),
    (10, 'E', 'F'),
    (2, 'F', 'G'),
    (1, 'G', 'H'), (6, 'G', 'I'),
    (7, 'H', 'I')
]

mst_kruskal = kruskals_mst(N, E)
print(f"\nKruskal's MST Edges: {mst_kruskal}")

File: c++/Graphs/kruskals_algorithm.cpp
// Problem: Kruskal's Algorithm - Minimum Spanning Tree (MST)
// Language: C++
// Topic: Graphs / Greedy Algorithms / Disjoint Set Union
// Description:
// Kruskal’s algorithm finds the Minimum Spanning Tree (MST) of a connected, weighted graph.
// It sorts all edges in ascending order of weight and adds them one by one,
// avoiding cycles using the Disjoint Set Union (Union-Find) data structure.
// Time Complexity: O(E * logE), where E = number of edges
// Space Complexity: O(V), where V = number of vertices

#include <bits/stdc++.h>
using namespace std;

class DSU {
public:
    vector<int> parent, rank;

    DSU(int n) {
        parent.resize(n);
        rank.resize(n, 0);
        for (int i = 0; i < n; i++)
            parent[i] = i;
    }

    int find(int x) {
        if (x == parent[x])
            return x;
        return parent[x] = find(parent[x]); // Path compression
    }

    bool unite(int x, int y) {
        int px = find(x);
        int py = find(y);

        if (px == py)
            return false; // Same component, adding this edge creates a cycle

        // Union by rank
        if (rank[px] < rank[py])
            parent[px] = py;
        else if (rank[py] < rank[px])
            parent[py] = px;
        else {
            parent[py] = px;
            rank[px]++;
        }

        return true;
    }
};

class Solution {
public:
    // Each edge is represented as {u, v, weight}
    int kruskalMST(int V, vector<vector<int>>& edges) {
        // Sort edges by weight
        sort(edges.begin(), edges.end(),
             [](const vector<int>& a, const vector<int>& b) {
                 return a[2] < b[2];
             });

        DSU dsu(V);
        int mstWeight = 0;
        vector<vector<int>> mstEdges;

        for (auto& e : edges) {
            int u = e[0], v = e[1], w = e[2];

            if (dsu.unite(u, v)) {
                mstWeight += w;
                mstEdges.push_back({u, v, w});
            }
        }

        // Print MST edges
        cout << "Edges in the Minimum Spanning Tree:\n";
        for (auto& e : mstEdges)
            cout << e[0] << " - " << e[1] << " : " << e[2] << "\n";

        return mstWeight;
    }
};

int main() {
    /*
        Example Graph:
        Vertices = 4
        Edges = 5
        Edge list: (u, v, w)
        0 -- 1 (10)
        0 -- 2 (6)
        0 -- 3 (5)
        1 -- 3 (15)
        2 -- 3 (4)

        Expected MST:
        Edges -> (2,3,4), (0,3,5), (0,1,10)
        Total Weight = 19
    */

    int V = 4;
    vector<vector<int>> edges = {
        {0, 1, 10},
        {0, 2, 6},
        {0, 3, 5},
        {1, 3, 15},
        {2, 3, 4}
    };

    Solution sol;
    int totalWeight = sol.kruskalMST(V, edges);

    cout << "\nTotal weight of MST: " << totalWeight << endl;

    return 0;
}
